### ISSUE TITLE
Fail to build: undefined reference to OMP_

### ISSUE THREAD

**Issue by DaniloZZZ** [ORNL-QCI/exatn#35]
Hi

I try to build exatn and run into an undefined reference error.
See the link to the gist for logs and all the versions in cmake

```
-- The CXX compiler identification is GNU 10.1.0
-- The Fortran compiler identification is GNU 10.1.0
```

```
⟩ uname -a
Linux archlinux 5.6.15-arch1-1 #1 SMP PREEMPT Wed, 27 May 2020 23:42:26 +0000 x86_64 GNU/Linux
```

https://gist.github.com/DaniloZZZ/e391f25e2c8a908d7f7fe533a2134de4

I do have mpi installed and looks like I'm having the issue described in #28 .

However, I have the same issue with the same errors on another Windows WSL machine (Ubuntu 20.04), which does not have MPI and Cmake doesn't find it.

**Comment by DmitryLyakh** [comment]
Confirmed, this is an issue in our current build system. It happens because a 3rd-party LAPACK library has become a mandatory dependence while a 3-rd party BLAS library is still optional. It will be fixed. But for now (and actually always) you should use an optimized BLAS library, for example OpenBLAS or Intel MKL. You can configure ExaTN with OpenBLAS like this (install OpenBLAS first and then provide the path to its lib directory below):
cmake .. -DEXATN_BUILD_TESTS=TRUE -DBLAS_LIB=OPENBLAS -DBLAS_PATH=/usr/local/blas/openblas/lib -DWITH_LAPACK=YES

**Comment by DaniloZZZ** [comment]
I'm building on my personal system, so I already have BLAS installed in `/usr/lib`. Running what you said results in a different issue concerning links to BLAS methods when compiling fortran files.

I found that in your `README.md` there is no option to use `BLAS_LIB=OPENBLAS`, and I tried to use `ATLAS` instead. as in your example configs.
```
 Example of a typical workstation configuration with default Linux BLAS (found in /usr/lib):
cmake ..
-DCMAKE_BUILD_TYPE=Release
-DEXATN_BUILD_TESTS=TRUE
-DBLAS_LIB=ATLAS -DBLAS_PATH=/usr/lib
```

This time it crashed later, at 79%, with the following error:

```
[ 79%] Building CXX object tpls/boost-cmake/CMakeFiles/Boost_regex.dir/boost/boost_1_64_0/libs/regex/src/c_regex_traits.cpp.o
cd /home/dali/git-bulit/exatn/build/tpls/boost-cmake && /usr/bin/c++  -DBOOST_ALL_NO_LIB=1 -DBOOST_DISABLE_ASSERT -DBOOST_SYSTEM_NO_DEPRECATED -DBOOST_THREAD_PROVIDES_EXECUTORS -DBOOST_THREAD_USES_CHRONO -DBOOST_THREAD_VERSION=4 -I/home/dali/git-bulit/exatn/build/tpls/cppmicroservices/include -I/home/dali/git-bulit/exatn/build/tpls/cppmicroservices/framework/include -I/home/dali/git-bulit/exatn/tpls/cppmicroservices/framework/include -isystem /home/dali/git-bulit/exatn/tpls/boost-cmake/boost/boost_1_64_0  -DPATH_MAX=4096 -Wno-attributes -DNO_GPU -DEXATN_SERVICE -Wno-error=deprecated-copy -O3 -DNDEBUG -fPIC   -w -std=gnu++14 -o CMakeFiles/Boost_regex.dir/boost/boost_1_64_0/libs/regex/src/c_regex_traits.cpp.o -c /home/dali/git-bulit/exatn/tpls/boost-cmake/boost/boost_1_64_0/libs/regex/src/c_regex_traits.cpp
In file included from /home/dali/git-bulit/exatn/tpls/boost-cmake/boost/boost_1_64_0/libs/regex/src/c_regex_traits.cpp:22:
/home/dali/git-bulit/exatn/tpls/boost-cmake/boost/boost_1_64_0/boost/config.hpp:35:12: fatal error: boost/config/select_compiler_config.hpp: No such file or directory
   35 | #  include <boost/config/select_compiler_config.hpp>
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I do have boost installed, and it's version  `1.72.0-2`. However, it looks like make tries to compile boost from source in tpln.

**Comment by DmitryLyakh** [comment]
ExaTN supports OPENBLAS (libopenblas.so) explicitly but you need to have it installed on your system, that is, you need to provide -DBLAS_PATH to where your libopenblas.so resides. ATLAS is a default Linux BLAS (libblas.so), which you can also use but it is not necessarily multi-threaded, that's why I recommended OpenBLAS (from github).
You do not need your system boost library, ExaTN has its own submodule. I think your system may have some issues with having multiple C++ compilers that confuse cmake. To re-confirm, did you specify your compilers explicitly, like:
CC=gcc CXX=g++ FC=gfortran cmake .. -DEXATN_BUILD_TESTS=TRUE -DBLAS_LIB=OPENBLAS -DBLAS_PATH=/usr/local/blas/openblas/lib -DWITH_LAPACK=YES
or, if you still want to use ATLAS,
CC=gcc CXX=g++ FC=gfortran cmake .. -DEXATN_BUILD_TESTS=TRUE -DBLAS_LIB=ATLAS -DBLAS_PATH=/usr/lib -DWITH_LAPACK=YES
This should help CMAKE make the right choice.

**Comment by DaniloZZZ** [comment]
I didn't include it previously, but I did re-run both `OPENBLAS` and `ATLAS` options with explicitly specified compilers and got the same results. 

I used the following:
```
⟩ CC=/usr/bin/gcc CXX=/usr/bin/g++ FC=/usr/bin/gfortran cmake .. -DEXATN_BUILD_TESTS=TRUE -DBLAS_LIB=ATLAS -DBLAS_PATH=/usr/lib -DWITH_LAPACK=YES
```

`ATLAS` fails with `fatal error: boost/config/select_compiler_config.hpp: No such file or directory`

`OPENBLAS` fails at 68% with following:

```
[ 68%] Built target antlr4_shared
make  -f tpls/CMakeFiles/exatensor-build.dir/build.make tpls/CMakeFiles/exatensor-build.dir/depend
make[2]: Entering directory '/home/dali/git-bulit/exatn/build'
cd /home/dali/git-bulit/exatn/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/dali/git-bulit/exatn /home/dali/git-bulit/exatn/tpls /home/dali/git-bulit/exatn/build /home/dali/git-bulit/exatn/build/tpls /home/dali/git-bulit/exatn/build/tpls/CMakeFiles/exatensor-build.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/dali/git-bulit/exatn/build'
make  -f tpls/CMakeFiles/exatensor-build.dir/build.make tpls/CMakeFiles/exatensor-build.dir/build
make[2]: Entering directory '/home/dali/git-bulit/exatn/build'
cd /home/dali/git-bulit/exatn/tpls/ExaTensor && /usr/bin/cmake -E env CMAKE_CXX_COMPILER=/opt/cuda/bin/g++ CMAKE_C_COMPILER=/opt/cuda/bin/gcc CMAKE_Fortran_COMPILER=/usr/bin/gfortran TOOLKIT=GNU BUILD_TYPE=OPT EXA_NO_BUILD=NO EXA_TALSH_ONLY=YES EXATN_SERVICE=YES EXA_OS=LINUX GPU_CUDA=NOCUDA MPILIB=NONE BLASLIB=OPENBLAS PATH_BLAS_OPENBLAS=/usr/lib PATH_INTEL= WITH_LAPACK=YES PATH_LAPACK_LIB=/usr/lib LAPACK_LIBS="" make
make[3]: Entering directory '/home/dali/git-bulit/exatn/tpls/ExaTensor'
make -C ./TALSH
make[4]: Entering directory '/home/dali/git-bulit/exatn/tpls/ExaTensor/TALSH'
/usr/bin/gfortran ./OBJ/main.o ./OBJ/test.o libtalsh.a -L. -L/usr/lib -lopenblas -L/usr/lib  -lgomp -L. -L. -lstdc++
-o test_talsh.x
/usr/bin/ld: libtalsh.a(tensor_algebra_cpu.o): in function `__tensor_algebra_cpu_MOD_tensor_block_decompose_svd':
tensor_algebra_cpu.F90:(.text+0x394cc): undefined reference to `sgesvdx_'
/usr/bin/ld: tensor_algebra_cpu.F90:(.text+0x39d0f): undefined reference to `sgesvdx_'
/usr/bin/ld: tensor_algebra_cpu.F90:(.text+0x3a330): undefined reference to `zgesvdx_'
/usr/bin/ld: tensor_algebra_cpu.F90:(.text+0x3a527): undefined reference to `cgesvdx_'
/usr/bin/ld: tensor_algebra_cpu.F90:(.text+0x3a6b6): undefined reference to `zgesvdx_'
/usr/bin/ld: tensor_algebra_cpu.F90:(.text+0x3a8d7): undefined reference to `cgesvdx_'
/usr/bin/ld: tensor_algebra_cpu.F90:(.text+0x3aea3): undefined reference to `dgesvdx_'
/usr/bin/ld: tensor_algebra_cpu.F90:(.text+0x3b02b): undefined reference to `dgesvdx_'
collect2: error: ld returned 1 exit status
```
My OpenBlas installation:

```bash
⟩ ls /usr/lib/libopenbla* -la
lrwxrwxrwx 1 root root       22 Mar  4 05:05 /usr/lib/libopenblas.so -> libopenblasp-r0.3.9.so*
lrwxrwxrwx 1 root root       22 Mar  4 05:05 /usr/lib/libopenblas.so.3 -> libopenblasp-r0.3.9.so*
-rwxr-xr-x 1 root root 18589168 Mar  4 05:05 /usr/lib/libopenblasp-r0.3.9.so*
```

**Comment by DmitryLyakh** [comment]
Concerning OpenBLAS there is a possibility that your default libopenblas.so was built without or with an incomplete LAPACK support. You can check that by "nm libopenblas.so | grep -i gesvdx". If you see nothing, that's the case. I would recommend building an up-to-date OpebBLAS yourself from https://github.com/xianyi/OpenBLAS.

Concerning C++ issues, I still see CMAKE is trying to use different compilers available on your system. You should delete your ExaTN directory and try a fresh install from scratch with additional parameters:
git clone --recursive https://github.com/ORNL-QCI/exatn.git
cd exatn
git submodule update --init --recursive
mkdir build
cd build
CC=/usr/bin/gcc CXX=/usr/bin/g++ FC=/usr/bin/gfortran cmake .. -DCMAKE_C_COMPILER=/usr/bin/gcc -DCMAKE_CXX_COMPILER=/usr/bin/g++ -DCMAKE_Fortran_COMPILER=/usr/bin/gfortran -DEXATN_BUILD_TESTS=TRUE -DBLAS_LIB=ATLAS -DBLAS_PATH=/usr/lib -DWITH_LAPACK=YES

**Comment by DaniloZZZ** [comment]
`nm libopenblas.so` errors with 'No symbols'. I tried `nm -D` and it showed a list of something, but there is no `gesvdx`. There is an openblas-lapack package in AUR, which I installed, then removed and re-downloaded everything and then **exatn finally installed**, with `ATLAS`. 

In my previous message, I pasted the wrong output, sorry. I did my first run with `CC=gcc CXX=g++ FG=gfortran`, but by default that resolves to the compiler in `/opt/cuda/` that are links to version 8 compilers. I did run with the /usr/bin/ compilers, and got the same output. I guess that issue was in a broken download, maybe I forgot --recursive flag when initing the repo.

Thank you!
